File: vespalib/src/vespa/vespalib/trace/slime_trace_serializer.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "slime_trace_serializer.h"
#include "trace.h"
#include <cassert>

using namespace vespalib::slime;

namespace vespalib {

const Memory SlimeTraceSerializer::TIMESTAMP("timestamp");
const Memory SlimeTraceSerializer::PAYLOAD("payload");
const Memory SlimeTraceSerializer::CHILDREN("children");

SlimeTraceSerializer::SlimeTraceSerializer(Cursor & cursor)
    : _cursors()
{
    _cursors.push(&cursor);
}

void
SlimeTraceSerializer::visit(const TraceNode & node)
{
    assert(!_cursors.empty());
    Cursor * current(_cursors.top());
    assert(current != NULL);
    _cursors.pop();
    addTimestamp(*current, node);
    addPayload(*current, node);
    addChildrenCursors(*current, node);
}

void
SlimeTraceSerializer::addTimestamp(Cursor & current, const TraceNode & node)
{
    current.setLong(TIMESTAMP, count_ms(node.getTimestamp().time_since_epoch()));
}

void
SlimeTraceSerializer::addPayload(Cursor & current, const TraceNode & node)
{
    if (node.hasNote()) {
        current.setString(PAYLOAD, Memory(node.getNote()));
    }
}

void
SlimeTraceSerializer::addChildrenCursors(Cursor & current, const TraceNode & node)
{
    if (node.getNumChildren() > 0) {
        addChildrenCursorsToStack(current.setArray(CHILDREN), node);
    }
}

void
SlimeTraceSerializer::addChildrenCursorsToStack(Cursor & childrenArray, const TraceNode & node)
{
    for (uint32_t childIndex(0); childIndex < node.getNumChildren(); childIndex++) {
        Cursor & child(childrenArray.addObject());
        _cursors.push(&child);
    }
}

}

File: config/src/main/java/com/yahoo/vespa/config/protocol/SlimeTraceSerializer.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.config.protocol;

import com.yahoo.slime.Cursor;
import com.yahoo.yolean.trace.TraceNode;
import com.yahoo.yolean.trace.TraceVisitor;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;

/**
 * Serialize a {@link TraceNode} to {@link com.yahoo.slime.Slime}.
 *
 * @author Ulf Lilleengen
 */
public class SlimeTraceSerializer extends TraceVisitor {
    static final String TIMESTAMP = "timestamp";
    static final String PAYLOAD = "payload";
    static final String CHILDREN = "children";
    final Deque<Cursor> cursors = new ArrayDeque<>();

    public SlimeTraceSerializer(Cursor cursor) {
        cursors.push(cursor);
    }

    @Override
    public void visit(TraceNode node) {
        Cursor current = cursors.pop();
        current.setLong(TIMESTAMP, node.timestamp());
        encodePayload(current, node.payload());
        addChildrenCursors(current, node);
    }

    private void encodePayload(Cursor current, Object payload) {
        if (payload instanceof String) {
            current.setString(PAYLOAD, (String)payload);
        } else if (payload instanceof Long) {
            current.setLong(PAYLOAD, (Long) payload);
        } else if (payload instanceof Boolean) {
            current.setBool(PAYLOAD, (Boolean) payload);
        } else if (payload instanceof Double) {
            current.setDouble(PAYLOAD, (Double) payload);
        } else if (payload instanceof byte[]) {
            current.setData(PAYLOAD, (byte[]) payload);
        }
    }

    private void addChildrenCursors(Cursor current, TraceNode node) {
        Iterator<TraceNode> it = node.children().iterator();
        if (it.hasNext()) {
            Cursor childrenArray = current.setArray(CHILDREN);
            while (it.hasNext()) {
                cursors.push(childrenArray.addObject());
                it.next();
            }
        }
    }
}
